Add instructions and config files for installing at ASU. (#2)
epic-astronomy/LWA_EPIC: INSTALL_ASU.md

# General

Start with the general [Enterprise new user setup](https://docs.google.com/document/d/1DOoiYEZd15KMw4KNSIkaGnph2aARI4n77cEuQHiOJqs/edit), and do the steps up to and including anaconda installation. When you create your environment, **be sure to use python 2**, not 3 as is the default. Give an appropriate name to the conda environment. Below we use `epic` as the example.

After the installation of anaconda do the following
```
$ user=`whoami`
$ env=epic
 $ conda create -n ${env} python=2.7
$ conda activate ${env}
$ conda install ipython
$ python -m ipykernel install --user --name ${env} --display-name ${env}
$ pip install lsl
```

# Bifrost

> Until some pull requests are merged, we need to use a hybrid of two bifrost forks.

Author: Note James' romein branch has been merged in! We can update this to just use LEDA bifrost, but as of writing, we still need to use master + plugin-wrapper because the latter hasn't been updated.

# General

Start with the general [Enterprise new user setup](https://docs.google.com/document/d/1DOoiYEZd15KMw4KNSIkaGnph2aARI4n77cEuQHiOJqs/edit), and do the steps up to and including anaconda installation. When you create your environment, **be sure to use python 2**, not 3 as is the default. Give an appropriate name to the conda environment. Below we use `epic` as the example.

After the installation of anaconda do the following
```
$ user=`whoami`
$ env=epic
 $ conda create -n ${env} python=2.7
```
It is advised to install the dependencies of LSL before attempting a `pip install`. While pip generally is good about grabbing package dependencies, the configuration of some packages requires dependencies to be installed before pip can check if the dependencies are installed.

While not explicitly mentioned, lsl depends on `astropy` for fits I/O. Inclusion of astropy is not a mistake.

To install dependencies for lsl excute the following lines. This may not be necessay depending on your environment but will help smooth the install.
```
$ conda activate ${env}

> $ conda install aipy scipy numpy fftw astropy healpy ephem pytz matplotlib

Author: healpy and aipy both require conda-forge. Should add line before this one:
```suggestion
$ conda config --add channels conda-forge
$ conda install aipy scipy numpy fftw astropy healpy ephem pytz matplotlib
```

Author: I also had to install atlas
```suggestion
$ conda install -c anaconda atlas
$ conda config --add channels conda-forge
$ conda install aipy scipy numpy fftw astropy healpy ephem pytz matplotlib
```

$ pip install lsl
```

# Bifrost

> Until some pull requests are merged, we need to use a hybrid of two bifrost forks.

Author: ```suggestion
```

```
$ cd ~/src
$ git clone https://github.com/ledatelescope/bifrost.git -b plugin-wrapper bifrost
$ cd bifrost 

> $ git merge master  # merge in the lastest master to get work from James Kent's fork

Author: We now have our own fork, which we'll try to keep up to date.
```suggestion
$ git clone https://github.com/epic-astronomy/bifrost.git
$ cd bifrost 
# For development work, you will want to checkout the appropriate branch. For example:
$ git checkout plugin-wrapper
```

$ cp /data4/jdowell/CodeSafe/bifrost/src/proclog.cpp src/  # Make multi-user installation work
```
(The `proclog.ccp` file referenced above is also in this repository, `LWA_EPIC/config/ASU_proclog.cpp`, so you can copy from there in case anything ever happens to the one on `/data4/`.)

> Check `README.md` to install dependencies (do this in your epic env).

Author: ```suggestion
Check `README.md` to install dependencies (do this in your epic env, *not* with sudo as the README indicates).
```

Reviewer: I think the `proclog.cpp` fix is in the main Bifrost repo now.

> Always check that the ctypesgen is the latest as mentioned in the README.md

Author: This comment has me confused. Why are we specifying a commit if we want the latest? And the line below points to a different commit than what's in the README, so I'm not sure which one I should be using.

Author: Just found a comment in the slack channel that says the one in the README doesn't work anymore. We still need to clarify which to use and why.

```
$ pip install numpy contextlib2 pint git+https://github.com/olsonse/ctypesgen.git@9bd2d249aa4011c6383a10890ec6f203d7b7990f
```

Downgrade matplotlib
```
$ conda install matplotlib=2.2.3
```

Now we need to actually install bifrost. First set up configuration for intrepid. Copy the file `LWA_EPIC/config/ASU_user.mk` from this repository to your bifrost directory:
```

> $ cp ~/src/LWA_EPIC/config/ASU_user.mk ~/bifrost/user.mk

Author: ```suggestion
$ cp ~/src/LWA_EPIC/config/ASU_user.mk ~/src/bifrost/user.mk
```

```

Finally, install (note **this must be done on intrepid with your environment activated**):
```
$ make -j 32
$ make install INSTALL_LIB_DIR="/home/${user}/src/anaconda/envs/${env}/lib" INSTALL_INC_DIR="/home/${user}/src/anaconda/envs/${env}/include" PYINSTALLFLAGS="--prefix=/home/${user}/src/anaconda/envs/${env}"
```

### Testing multiple branches
If you want to install and test multiple bifrost branches, it is encouraged to use different conda environments. Check the environment variables of the install directory to properly link with the libraries.

# EPIC
https://github.com/epic-astronomy/epic

> TODO: Finish EPIC install instructions.

Author: I don't think this is needed anymore. But we probably do want to add an example showing how to run the script. And maybe even an example image to confirm you get the right thing.

# General

Start with the general [Enterprise new user setup](https://docs.google.com/document/d/1DOoiYEZd15KMw4KNSIkaGnph2aARI4n77cEuQHiOJqs/edit), and do the steps up to and including anaconda installation. When you create your environment, **be sure to use python 2**, not 3 as is the default. Give an appropriate name to the conda environment. Below we use `epic` as the example.

After the installation of anaconda do the following
```
$ user=`whoami`
$ env=epic
 $ conda create -n ${env} python=2.7
```
It is advised to install the dependencies of LSL before attempting a `pip install`. While pip generally is good about grabbing package dependencies, the configuration of some packages requires dependencies to be installed before pip can check if the dependencies are installed.

While not explicitly mentioned, lsl depends on `astropy` for fits I/O. Inclusion of astropy is not a mistake.

To install dependencies for lsl excute the following lines. This may not be necessay depending on your environment but will help smooth the install.
```
$ conda activate ${env}

> $ conda config --add channels conda-forge

Author: ```suggestion
$ conda install -c anaconda atlas
$ conda config --add channels conda-forge
```

Reviewer: I'll go ahead and throw this out there:  do you want to switch to the "new" LSL?

Author: sure? How does one do that?

Reviewer: It should be as easy as `pip install git+https://github.com/lwa-project/lsl.git` as long as you have libfftw3.  The caveats are: 
 1.  it has an incompatible API with the current LSL that LWA_EPIC uses (but there's a branch for that) and
2. it's technically not released yet.

Author: Ok, I think since there's an API change we should update the install instructions with the branch you're working on.

# General

Start with the general [Enterprise new user setup](https://docs.google.com/document/d/1DOoiYEZd15KMw4KNSIkaGnph2aARI4n77cEuQHiOJqs/edit), and do the steps up to and including anaconda installation. When you create your environment, **be sure to use python 2**, not 3 as is the default. Give an appropriate name to the conda environment. Below we use `epic` as the example.

After the installation of anaconda do the following
```
$ user=`whoami`
$ env=epic
 $ conda create -n ${env} python=2.7
```
It is advised to install the dependencies of LSL before attempting a `pip install`. While pip generally is good about grabbing package dependencies, the configuration of some packages requires dependencies to be installed before pip can check if the dependencies are installed.

While not explicitly mentioned, lsl depends on `astropy` for fits I/O. Inclusion of astropy is not a mistake.

To install dependencies for lsl excute the following lines. This may not be necessay depending on your environment but will help smooth the install.
```
$ conda activate ${env}

> $ conda install -c anaconda atlas

Author: ```suggestion
$ conda install -c free atlas
```

$ conda config --add channels conda-forge
$ conda install aipy scipy numpy fftw astropy healpy ephem pytz matplotlib
$ pip install lsl
```

# Bifrost
```
$ cd ~/src
$ git clone https://github.com/epic-astronomy/bifrost.git

> $ cd bifrost 

Author: ```suggestion
$ git clone https://github.com/epic-astronomy/bifrost.git
# You will also need this repository. Clone if you haven't already
$ git clone https://github.com/epic-astronomy/LWA_EPIC.git
$ cd bifrost 
```

> # For development work, you will want to checkout the appropriate branch. For example:
> $ git checkout plugin-wrapper
> $ cp /data4/jdowell/CodeSafe/bifrost/src/proclog.cpp src/  # Make multi-user installation work

Author: These lines make it look like the proclog bit is only relevant for development work.
```suggestion
$ cp /data4/jdowell/CodeSafe/bifrost/src/proclog.cpp src/  # Make multi-user installation work
# For development work, you will want to checkout the appropriate branch. For example:
$ git checkout plugin-wrapper
```

> # For development work, you will want to checkout the appropriate branch. For example:
> $ git checkout plugin-wrapper
> $ cp /data4/jdowell/CodeSafe/bifrost/src/proclog.cpp src/  # Make multi-user installation work

Author: These lines make it look like the proclog bit is only relevant for development work.
```suggestion
$ cp ~/src/LWA_EPIC/config/ASU_proclog.cpp src/proclog.cpp  # Make multi-user installation work
# For development work, you will want to checkout the appropriate branch. For example:
$ git checkout plugin-wrapper
```

```

> (The `proclog.ccp` file referenced above is also in this repository, `LWA_EPIC/config/ASU_proclog.cpp`, so you can copy from there in case anything ever happens to the one on `/data4/`.)

Author: Update to just use this repository. Makes this comment unnecessary.
```suggestion
```

While not explicitly mentioned, lsl depends on `astropy` for fits I/O. Inclusion of astropy is not a mistake.

To install dependencies for lsl excute the following lines. This may not be necessay depending on your environment but will help smooth the install.
```
$ conda activate ${env}
$ conda install -c anaconda atlas
$ conda config --add channels conda-forge
$ conda install aipy scipy numpy fftw astropy healpy ephem pytz matplotlib
$ pip install lsl
```

# Bifrost
```
$ cd ~/src
$ git clone https://github.com/epic-astronomy/bifrost.git
$ cd bifrost 
# For development work, you will want to checkout the appropriate branch. For example:
$ git checkout plugin-wrapper
$ cp /data4/jdowell/CodeSafe/bifrost/src/proclog.cpp src/  # Make multi-user installation work
```
(The `proclog.ccp` file referenced above is also in this repository, `LWA_EPIC/config/ASU_proclog.cpp`, so you can copy from there in case anything ever happens to the one on `/data4/`.)

> Check `README.md` to install dependencies (do this in your epic env, *not* with sudo as the README indicates).
> Always check that the ctypesgen is the latest as mentioned in the README.md

Author: ```suggestion
Note that the `README.md` in bifrost can be (and as of this commit *is*) out of date. Use the following line to install dependencies. (Do this in your epic env, *not* with sudo as the bifrost README indicates). The ctypes gen commit referenced here works for Ubuntu >18. In the future this may be another point of trouble. If this version doesn't work, you will want to file and issue on the [ledatelescope bifrost repo](https://github.com/ledatelescope/bifrost/issues).
```

```

> $ pip install numpy contextlib2 pint git+https://github.com/olsonse/ctypesgen.git@9bd2d249aa4011c6383a10890ec6f203d7b7990f

Reviewer 2: why is `numpy` listed in this line and also in line 21

Author: ```suggestion
$ pip install contextlib2 pint git+https://github.com/olsonse/ctypesgen.git@9bd2d249aa4011c6383a10890ec6f203d7b7990f
```

```

Downgrade matplotlib
```
$ conda install matplotlib=2.2.3
```

Now we need to actually install bifrost. First set up configuration for intrepid. Copy the file `LWA_EPIC/config/ASU_user.mk` from this repository to your bifrost directory:
```
$ cp ~/src/LWA_EPIC/config/ASU_user.mk ~/src/bifrost/user.mk
```

Finally, install (note **this must be done on intrepid with your environment activated**):
```
$ make -j 32
$ make install INSTALL_LIB_DIR="/home/${user}/src/anaconda/envs/${env}/lib" INSTALL_INC_DIR="/home/${user}/src/anaconda/envs/${env}/include" PYINSTALLFLAGS="--prefix=/home/${user}/src/anaconda/envs/${env}"
```

### Testing multiple branches
If you want to install and test multiple bifrost branches, it is encouraged to use different conda environments. Check the environment variables of the install directory to properly link with the libraries.

# Test Run
Remember to check that you are logged onto intrepid and in your environment.
```

> $ cd ~/src/LWA_EPIC/scripts

Author: ```suggestion
$ cd ~/src/LWA_EPIC/LWA
```

After the installation of anaconda do the following
```
$ user=`whoami`
$ env=epic
 $ conda create -n ${env} python=2.7
```
It is advised to install the dependencies of LSL before attempting a `pip install`. While pip generally is good about grabbing package dependencies, the configuration of some packages requires dependencies to be installed before pip can check if the dependencies are installed.

While not explicitly mentioned, lsl depends on `astropy` for fits I/O. Inclusion of astropy is not a mistake.

To install dependencies for lsl excute the following lines. This may not be necessay depending on your environment but will help smooth the install.
```
$ conda activate ${env}
$ conda install -c free atlas
$ conda config --add channels conda-forge
$ conda install aipy scipy numpy fftw astropy healpy ephem pytz matplotlib
$ pip install lsl
```

# Bifrost
```
$ cd ~/src
$ git clone https://github.com/epic-astronomy/bifrost.git
$ cd bifrost

> $ cp /data4/jdowell/CodeSafe/bifrost/src/proclog.cpp src/  # Make multi-user installation work

Author: ```suggestion
$ cp ~/src/LWA_EPIC/config/ASU_proclog.cpp src/proclog.cpp  # Make multi-user installation work
```
